src/adaptix/_internal/morphing/generic_provider.py

```diff
@@ -99,19 +99,61 @@ def _get_allowed_values_collection(self, args: Collection) -> Collection:
             return set(args)
         return tuple(args)
 
+    def _get_enum_loader(
+        self, mediator: Mediator, request: LoaderRequest, enum_class: Type[Enum]
+    ) -> Callable[[Any], Enum]:
+        return mediator.mandatory_provide(
+            LoaderRequest(
+                loc_stack=request.loc_stack.append_with(
+                    LocMap(
+                        TypeHintLoc(type=enum_class),
+                    )
+                )
+            ),
+            lambda x: f'Cannot create loader for {enum_class}. Loader for literal cannot be created',
+        )
+
+    def _get_enum_dumper(
+        self, mediator: Mediator, request: DumperRequest, enum_class: Type[Enum]
+    ) -> Callable[[Any], Enum]:
+        return mediator.mandatory_provide(
+            DumperRequest(
+                loc_stack=request.loc_stack.append_with(
+                    LocMap(
+                        TypeHintLoc(type=enum_class),
+                    )
+                )
+            ),
+            lambda x: f'Cannot create dumper for {enum_class}. Dumper for literal cannot be created',
+        )
+
+    def _combined_enums_loader(self, data, loaders: Collection, allowed_values: Collection) -> Optional[Enum]:
+        for loader in loaders:
+            try:
+                result = loader(data)
+                if result in allowed_values:
+                    return result
+            except BadVariantError:
+                pass
+        return None
+
+    def _with_enum_loader(self, func: Callable, loaders: Collection, allowed_values: Collection):
+        def wrapped(data):
+            enum_data = self._combined_enums_loader(data, loaders, allowed_values)
+            if enum_data:
+                return enum_data
+            return func(data)
+        return wrapped
+
     def _provide_loader(self, mediator: Mediator, request: LoaderRequest) -> Loader:
```

Comment:
We need to raise ``BadVariantError`` with appropriate list of allowed values when enum is using.
I think that dumping every allowed enum instance is quite ok for this situation.